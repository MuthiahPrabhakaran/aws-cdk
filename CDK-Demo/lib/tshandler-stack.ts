import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 } from 'aws-cdk-lib';
import { Runtime, Code, Function as LambdaFunction } from  'aws-cdk-lib/aws-lambda';


interface TsHandlerStackProps extends cdk.StackProps {
    coolBucket: aws_s3.Bucket
}

export class TsHandlerStack extends cdk.Stack {

    constructor(scope: Construct, id: string, props: TsHandlerStackProps) {
        super(scope, id, props);
        new LambdaFunction(this, 'PhotosHandler', {
            runtime: Runtime.NODEJS_18_X,
            handler: 'index.handler',
            code: Code.fromInline(`
                exports.handler = async (event) => {
                console.log("Event received:", event);
                return {
                  statusCode: 200,
                 body: JSON.stringify({ message: "Hello from inline Node.js Lambda!. Bucket name: " + process.env.COOL_BUCKET_ARN }),
                    };
                 };
                `),
                environment: {
                    COOL_BUCKET_ARN: props.coolBucket.bucketArn
                }
        })
    }

}