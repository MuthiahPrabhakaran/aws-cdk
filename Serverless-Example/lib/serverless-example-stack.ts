import * as cdk from 'aws-cdk-lib';
import { Construct, Node } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs'
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { join } from 'path';
import { LambdaIntegration, ResourceOptions, RestApi, Cors } from 'aws-cdk-lib/aws-apigateway';
import { AttributeType, Billing, TableV2 } from 'aws-cdk-lib/aws-dynamodb';

export class ServerlessExampleStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // DynamoDB Table
    const employeesTable = new TableV2(this, 'EmplTable', {
      partitionKey: {
          name: 'id',
          type: AttributeType.STRING,
        },
      billing: Billing.onDemand()
    });


    // Lambda
    const emplLambda = new NodejsFunction(this, 'Ts-EmplLambda', {
        runtime: Runtime.NODEJS_18_X,
        handler: 'handler',
        entry: (join(__dirname, '..', 'services', 'handler.ts')),
        environment: {
          TABLE_NAME: employeesTable.tableName
        }
  })

  // Permission to lambda to perform operation on DynamoDB
  employeesTable.grantReadWriteData(emplLambda);
  

  // API Gateway
  const api = new RestApi(this, 'TS-EmplApi');
  const optionsWithCors: ResourceOptions = {
    defaultCorsPreflightOptions: {
      allowOrigins: Cors.ALL_ORIGINS,
      allowMethods: Cors.ALL_METHODS
    }
  }
  const emplResource = api.root.addResource('empl', optionsWithCors);

  const emplLambdaIntegration = new LambdaIntegration(emplLambda);
  emplResource.addMethod('GET', emplLambdaIntegration);
  emplResource.addMethod('POST', emplLambdaIntegration);
  }
}
