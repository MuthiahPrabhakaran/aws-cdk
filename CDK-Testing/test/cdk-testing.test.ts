import * as cdk from 'aws-cdk-lib';
import { Template, Match, Capture } from 'aws-cdk-lib/assertions';
import * as CdkTesting from '../lib/cdk-testing-stack';
import { PolicyDocument } from 'aws-cdk-lib/aws-iam';

describe('CdkTestingStack test suite', () => {

    let template: cdk.assertions.Template;

    beforeAll(() => {
        const app = new cdk.App({
            outdir: 'cdk.out/test'
        });
        const stack = new CdkTesting.CdkTestingStack(app, 'MyTestStack');
        template = Template.fromStack(stack);

    })


    test('Lambda runtime check', () => {
        template.hasResourceProperties('AWS::Lambda::Function', {
            Runtime: "nodejs18.x"
        });
        template.resourceCountIs('AWS::Lambda::Function', 1);
    });

    test('Lambda runtime check', () => {
        template.hasResourceProperties('AWS::Lambda::Function', {
            Runtime: Match.stringLikeRegexp('nodejs')
        });
    });

    test('Lambda bucket policy with matchers', () => {
        template.hasResourceProperties('AWS::IAM::Policy',
            Match.objectLike({
                PolicyDocument: {
                    Statement: [{
                        Resource: [{
                            'Fn::GetAtt': [
                                Match.stringLikeRegexp('CdkTestingBucket'),
                                'Arn'
                            ]
                        },
                        Match.anyValue() // Because the actual statement contains Fn get and join. Since we test get, for join -> i am using any
                        ]
                    }]
                }
            })
        );
    });

    test('Lambda actions with capture', () => {
        const lambdaActionsCapture = new Capture();
        template.hasResourceProperties('AWS::IAM::Policy', { // after this gets executed, the capture object contains the actual object
            PolicyDocument: {
                Statement: [{
                    Action: lambdaActionsCapture
                }]
            }
        });
        const executedActions  = [
        "s3:GetObject*",
        "s3:GetBucket*",
        "s3:List*"
       ]
       expect(lambdaActionsCapture.asArray()).toEqual(
        expect.arrayContaining(executedActions)
       )
    });

    test('Bucket properties with snapshot', () => {
        const bucketTemplate = template.findResources("AWS::S3::Bucket");
        expect(bucketTemplate).toMatchSnapshot()
    })


});
