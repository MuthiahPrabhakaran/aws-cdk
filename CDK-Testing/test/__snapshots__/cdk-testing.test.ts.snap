// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdkTestingStack test suite Bucket properties with snapshot 1`] = `
{
  "CdkTestingBucket2C26D58A": {
    "DeletionPolicy": "Retain",
    "Type": "AWS::S3::Bucket",
    "UpdateReplacePolicy": "Retain",
  },
}
`;
